packages/linker/test/handler.test.ts

```diff
@@ -1,97 +1,140 @@
+    fixture.set(`<a href="${location.href}">${random.word()}</a>`, true)
 
-    assert(`${comUrl}?${agent.get('linkerParam')}` === a.href)
+    const a = document.getElementsByTagName('a')[0]
+    link([location.hostname], param, 10)({ target: a } as any)
+    assert(location.href === a.href)
   })
 
-  it('not match domain', () => {
-    const url = setUpLinkEvent(comUrl, ['example.net'])
-    assert(comUrl === url)
+  it('is defined query', () => {
+    // FIXME https://github.com/Microsoft/TypeScript/issues/21943
+    const l: any = toLink(`${internet.url()}?a=b#anker`)
+    fixture.set(`<a href="${l.href}" />`, true)
+    const a = document.getElementsByTagName('a')[0]
+    link([l.hostname], param, 10)({ target: a } as any)
+    assert(`${l.origin}${l.pathname}${l.search}&${param}${l.hash}` === a.href)
   })
 
-  it('not have href', () => {
-    const url = setUpLinkEvent(undefined, ['example.net'])
-    assert(url === '')
+  it('bubbling', () => {
+    const l = toLink(internet.url())
+    fixture.set(`<a href="${l.href}"><img src="${image.imageUrl()}"></a>`)
+
+    link([l.hostname], param, 10)({
+      target: document.getElementsByTagName('img')[0]
+    } as any)
+    assert(`${l.href}?${param}` === document.getElementsByTagName('a')[0].href)
+
+    fixture.cleanup()
+    fixture.set(
+      `<a href="${l.href}"><div><img src="${image.imageUrl()}"></div></a>`
+    )
+
+    link([l.hostname], param, 1)({
+      target: document.getElementsByTagName('img')[0]
+    } as any)
+    assert(
+      l.href === document.getElementsByTagName('a')[0].href,
+      'already checked max parentNode'
+    )
   })
 
   it('submit post', () => {
-    const handler = submit([domain], agent.get('linkerParam'))
-    const form = createForm(comUrl, 'post')
-    handler({ target: form } as any)
-    assert(`${comUrl}?${agent.get('linkerParam')}` === form.action)
+    const l = toLink(internet.url())
+    fixture.set(`<form method="post" action="${l.href}"><form>`, true)
+
+    const form = document.getElementsByTagName('form')[0]
+    submit([l.hostname], param)({ target: form } as any)
+    assert(`${l.href}?${param}` === form.action)
+  })
+
+  it('submit javascript:void(0);', () => {
+    const l = toLink(internet.url())
+    const action = 'javascript:void(0)'
+    fixture.set(
+      `<form action="${action}"><input type="submit" value="test" ></form>`,
+      true
+    )
+
+    const form = document.getElementsByTagName('form')[0]
+    submit([l.hostname], param)({ target: form } as any)
+    assert(action === form.action)
   })
 
   it('not cross domain', () => {
-    const url = document.location.href
-    const handler = submit([domain], agent.get('linkerParam'))
-    const form = createForm(document.location.href, 'post')
-    handler({ target: form } as any)
-    assert(url === form.action, 'not added query string')
+    const l = toLink(internet.url())
+    fixture.set(`<form method="get" action="${l.href}"><form>`, true)
```

Comment:
query params will not add if submit `method='get'` form.

Reply:
see https://github.com/userdive/agent.js/pull/368/files#diff-fa2b92802143c2dc45119226f16189cbR120

Reply:
Yes, it is good and working. But, i think this is not `not cross domain` test case.